ckan/tests/factories.py

```diff
@@ -401,6 +401,32 @@ def _create(cls, target_class, *args, **kwargs):
         return obj
 
 
+class APIToken(factory.Factory):
+    """A factory class for creating CKAN API Tokens"""
+
+    class Meta:
+        model = ckan.model.ApiToken
+
+    name = "first token"
+    user = factory.LazyAttribute(lambda _: User()["name"])
+
+    @classmethod
+    def _build(cls, target_class, *args, **kwargs):
+        raise NotImplementedError(".build() isn't supported in CKAN")
+
+    @classmethod
+    def _create(cls, target_class, *args, **kwargs):
+        if args:
+            assert False, "Positional args aren't supported, use keyword args."
+
+        context = {"user": kwargs["user"]}
```

Comment:
Actually that func gives me TypeError.. that's why i have to pass it explicitly.. but i can make another helper func (e.g `def _get_user_from_kwargs(kwargs):` ) to make this work

https://github.com/ckan/ckan/blob/604e86ddf055a6c457d6d8a0c59deb1b4ddd9542/ckan/tests/factories.py#L65-L66
This is where i get the TypeError.

Thanks

Reply:
It gives you `TypeError` because you should pass the full user dict as the `user` param in a factory, not just the name, eg change this:

```
user_token = factories.APIToken(user=user["name"])
```

with:

```
user_token = factories.APIToken(user=user)
```

(I actually think this is confusing and we should match the `context['user']` behaviour, but it's the current behaviour so let's keep it like that)